### ISSUE TITLE
Suboptimal names in 0.2.0

### ISSUE THREAD

**Issue by MahdiBM** [apple/swift-openapi-generator#244]
There seems to be too many `hyphen`s added to OpenAPI-generated-code in [Penny](https://github.com/vapor/penny-bot/tree/acd48c7b9897ea1125a1543577aa2dab54412568/Lambdas/GitHubAPI)
Current code is not fully updated with `0.2.0`. You can clone the repo, resolve package versions and rerun the command-plugin if you need to test with Penny.
```diff
public typealias Repository = Components.Schemas.repository
- public typealias User = Components.Schemas.simple_user
- public typealias Organization = Components.Schemas.organization_simple
- public typealias PullRequest = Components.Schemas.pull_request
- public typealias SimplePullRequest = Components.Schemas.pull_request_simple
+ public typealias User = Components.Schemas.simple_hyphen_user
+ public typealias Organization = Components.Schemas.organization_hyphen_simple
+ public typealias PullRequest = Components.Schemas.pull_hyphen_request
+ public typealias SimplePullRequest = Components.Schemas.pull_hyphen_request_hyphen_simple
public typealias Issue = Components.Schemas.issue
public typealias Label = Components.Schemas.label
- public typealias InstallationToken = Components.Schemas.installation_token
- public typealias NullableUser = Components.Schemas.nullable_simple_user
+ public typealias InstallationToken = Components.Schemas.installation_hyphen_token
+ public typealias NullableUser = Components.Schemas.nullable_hyphen_simple_hyphen_user
public typealias Contributor = Components.Schemas.contributor
- public typealias PullRequestReviewComment = Components.Schemas.pull_request_review_comment
- public typealias DiffEntry = Components.Schemas.diff_entry
+public typealias PullRequestReviewComment = Components.Schemas.pull_hyphen_request_hyphen_review_hyphen_comment
+ public typealias DiffEntry = Components.Schemas.diff_hyphen_entry
public typealias Commit = Components.Schemas.commit
- public typealias SimpleCommit = Components.Schemas.simple_commit
- public typealias Committer = Components.Schemas.simple_commit.committerPayload
+ public typealias SimpleCommit = Components.Schemas.simple_hyphen_commit
+ public typealias Committer = SimpleCommit.committerPayload
- public typealias Installation = Components.Schemas.simple_installation
+ public typealias Installation = Components.Schemas.simple_hyphen_installation
```
This also affects stuff like functions names:
```diff
- try await context.githubClient.repos_list_pull_requests_associated_with_commit(...
+ try await context.githubClient.repos_sol_list_hyphen_pull_hyphen_requests_hyphen_associated_hyphen_with_hyphen_commit(...
```

**Comment by czechboy0** [comment]
Hi @MahdiBM, this was the subject of proposal [SOAR-0001](https://swiftpackageindex.com/apple/swift-openapi-generator/main/documentation/swift-openapi-generator/soar-0001) to address conflicts of names when characters not legal in Swift identifiers are used.

A good way to have the names exactly what you want them to be is to only use characters legal in Swift identifiers in your OpenAPI document. That is, `a-zA-Z0-9_`. Any other characters get turned into deterministic and unique codes as you see above.

Closing as this is expected behavior.

**Comment by MahdiBM** [comment]
This is GitHub's OpenAPI document though, I don't think I can change that without non-trivial changes.

Part of a proposal or not, i would say this behavior should be changed in someway, just because it makes the openapi file of GutHub look like this.

**Comment by czechboy0** [comment]
Please review the [goals](https://swiftpackageindex.com/apple/swift-openapi-generator/0.2.0/documentation/swift-openapi-generator/project-scope-and-goals) of the project, which includes respecting the identifiers in the OpenAPI document, as opposed to trying to "prettify" them, which leads to name conflicts where there are non in the OpenAPI document.

One example is GitHub's `+1` and `-1` names, which do not represent valid Swift identifiers, so we had to come up with a scheme that produces deterministic identifiers, which don't conflict.

A few options to apply your preferences to how the generated code is used or looks:
1. Pre-process the OpenAPI document using custom logic to reflect your preferred names (for the things that don't affect the wire format). For example, this works well for schema names and operation identifiers. It cannot be applied to property names, which affect the coding keys and the deserialization, those have to be unchanged.
2. Wrap the generated client in your own library, which vends the types and functions in the names you prefer.

Check out the discussion in SOAR-0001, as our goal is to support the full breadth of the OpenAPI specification, we practically cannot process the identifiers from OpenAPI without risking conflicts. In this project, that is considered more important than the conciseness of generated code.

None of this would be a problem if OpenAPI only allowed `a-zA-Z0-9_` in all identifiers, but alas, it allows more characters, and the generated has to deal with them somehow, in a way that doesn't produce conflicts.

**Comment by czechboy0** [comment]
It seems that you're already using typealiases, so I hope this change shouldn't affect your codebase much? As you discovered, typealiases are another good way to use whatever names you prefer for the generated types.

**Comment by MahdiBM** [comment]
About typealiases, yeah they will help. But I'll still need to change those an also change the function calls.
A big deal, but not a deal breaker. The function calls will still look bad.

**Comment by czechboy0** [comment]
The function calls you can change, that's [`operationId`](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#fixed-fields-8) from the OpenAPI specification. It doesn't impact the wire format, so you can update those in your copy of the OpenAPI document.

From the specification:

> Unique string used to identify the operation. The id MUST be unique among all operations described in the API. The operationId value is case-sensitive. Tools and libraries MAY use the operationId to uniquely identify an operation, therefore, it is RECOMMENDED to follow common programming naming conventions.

Unfortunately, GitHub didn't follow the recommendation of making the names code-friendly with `foo/bar-baz`, you could just update them to like `foo_bar_baz` or `fooBarBaz`, whichever you prefer, and that'll be the exact function name you'll get in Swift.

**Comment by MahdiBM** [comment]
Trying to update Penny, I once again bumped into this problem.
I wonder if the problem is common enough to be worth trying to address 🤔
Changing GitHub's OpenAPI document is not feasible, there are way too many hyphens. 
Only other option remaining is to just endure the names.

**Comment by czechboy0** [comment]
I find typealiases are a quick and easy way to make the names more friendly, and for larger projects wrapping the generated API in your own layer (which is helpful for testing and layer separation anyway).

**Comment by MahdiBM** [comment]
Typealiases only solve a part of the problem. There are other things like function names which are much more cumbersome to "alias". This just results in way too porr of a user exprience for any kind of these openapi documents.

There could be some settings to e.g. "replace hyphens with empty-string/dash unless it causes a name conflict" that'll probably be best of both worlds.

**Comment by MahdiBM** [comment]
Testing-wise we haven't had a need to do such things as we just use a mocked `ClientTransport`.

**Comment by czechboy0** [comment]
"Unless it causes a conflict" is surprisingly difficult, trust me I tried, it'd basically require a rewrite of the generator to keep a global, stateful registry of types and the order in which you'd generate types would affect the names you get. Seemed to us like a too high of a price to pay. Complicating the generator just to try to subjectively prettify names is explicitly out of scope of the project: https://swiftpackageindex.com/apple/swift-openapi-generator/0.3.1/documentation/swift-openapi-generator/project-scope-and-goals

The proposal SOAR-0001 is meant to be the solution to how the generator processes names, and there'd have to be a very good new design that isn't going to overly complicate the generator (such as by keeping a global type registry), which would continue to accept all the OpenAPI documents we accept today, for us to consider changing the naming logic. But since it's also very subjective, I feel simply respecting the OpenAPI document is the most defensible position, as every other solution would a) produce conflicts or overly complicate the generator and b) be subject to bikeshedding, since "pretty names" is not a universally agreed RFC, so we'd likely spend a ton of time just on agreeing on the naming logic itself.

As documented here, if you care about a vended API, wrap it in a hand-written one: https://swiftpackageindex.com/apple/swift-openapi-generator/0.3.1/documentation/swift-openapi-generator/api-stability-of-generated-code

**Comment by MahdiBM** [comment]
A rather easy solution occured to me: find and replace "_hyphen_" with "_". Solves all problems. Works like a charm!

We already do mess around with the GitHub OpenAPI documentation so I can't say a find-and-replace like this is "cumbersome" compared to what we already do, although it makes Xcode lag a decent amount (next time will just use VSCode as is the case already for a bunch of other Swift-related activities. It's (basically) already better for anything other than _writing_ code, and only in _Swift_, IMO).

I still think this should be something to improve on. Maybe there could be a long term plan, or an issue about it being something at the back of your mind. Maybe someday someone finds a proper way to do it.

**Comment by czechboy0** [comment]
It won't work like a charm. We used to (pre-SOAR-0001) replace all illegal characters with an underscore.

GitHub failed to compile, because it has both "+1" and "-1", which we both turned into "_1".

See the discussions on the forums about SOAR-0001. This is surprisingly non-trivial, and we spent weeks exploring it. At this point, we'd need to see some new information to change course, but based on the information available when SOAR-0001 was in review, the current behavior is the best overall solution.

**Comment by MahdiBM** [comment]
O I just meant in my specific case it solves my problem, otherwise I know the generator itself can't be this careless and just do a find-and-replace 😅

**Comment by czechboy0** [comment]
Ah gotcha. In your case you could instead find and replace the operationId in the OpenAPI document to give it names that don't require any escaping.

**Comment by czechboy0** [comment]
Landed in 1.6.0 as the new idiomatic naming strategy: https://swiftpackageindex.com/apple/swift-openapi-generator/main/documentation/swift-openapi-generator/soar-0013
